### PR TITLE
app-eselect/eselect-blas: new design for BLAS/CBLAS runtime switch.

### PR DESCRIPTION
**Pull request by cdluminate**
Related discussion could be found on gentoo-dev. @heroxbd

### CODE REVIEW

**Comment by cdluminate** [comment]
@gentoo-bot I signed it!

File: app-eselect/eselect-blas/eselect-blas-0.2.ebuild [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Copyright 1999-2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
```
Comment (heroxbd) [inline app-eselect/eselect-blas/eselect-blas-0.2.ebuild]: Is there a reason not using the newest EAPI-7?

File: app-eselect/eselect-blas/files/blas.eselect-0.2 [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		[[ -r ${cand} ]] && basename ${cand}
+	done
+}
+
+### helper utility to validate a BLAS/CBLAS library directory
+validate_blaslibdir() {
+	[[ -d $1 ]] || return 2  # dir doesn't exist
+	[[ -r "${1}/libblas.so.3" ]] && [[ -r "${1}/libblas.so" ]] \
+	&& [[ -r "${1}/libcblas.so.3" ]] && [[ -r "${1}/libcblas.so" ]] \
+	&& return 0 || return 1
```
Comment (heroxbd) [inline app-eselect/eselect-blas/files/blas.eselect-0.2]: Is `return 0` needed?

File: app-eselect/eselect-blas/files/blas.eselect-0.2 [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	store_config ${EROOT}/etc/env.d/blas/${libdir}/config CURRENT "${impl}"
+	# generate ld.so.conf file
+	ldconf="${EROOT}/etc/ld.so.conf.d/81-blas-${libdir}.conf"
+	libpath=$(load_config ${EROOT}/etc/env.d/blas/${libdir}/${impl} LDPATH)
+	truncate -s0 $ldconf
+	echo "# Auto-generated by eselect::blas. DON'T EDIT." >> $ldconf
+	echo "$libpath" >> $ldconf
+
+	# refresh cache
+	ldconfig; env-update
```
Comment (heroxbd) [inline app-eselect/eselect-blas/files/blas.eselect-0.2]: As far as I know, env-update calls ldconfig. It's redundant.

**Comment by heroxbd** [comment]
Please also mention the Bug you are trying to solve, in the format of https://github.com/gentoo/gentoo/commit/f5c71a4aa9615c8c14caa8e7076519ab2a4b9824 for example.

**Comment by cdluminate** [comment]
@heroxbd Done.

**Comment by cdluminate** [comment]
```patch
diff --git a/app-eselect/eselect-blas/files/blas.eselect-0.2 b/app-eselect/eselect-blas/files/blas.eselect-0.2
index 83fd64ee818..001ddb941e8 100644
--- a/app-eselect/eselect-blas/files/blas.eselect-0.2
+++ b/app-eselect/eselect-blas/files/blas.eselect-0.2
@@ -128,7 +128,13 @@ do_set() {
    # validate arguments
    has "$libdir" "${libdirs[@]}" || die -q "Invalid libdir ($libdir) !"
    targets=( $(find_targets_in $libdir) )
-   has "$impl" "${targets[@]}" || die -q "Invalid impl ($impl) !"
+   if ! has "$impl" "${targets[@]}"; then
+       if [[ ${impl} -gt 0 ]] && [[ ${impl} -le ${#targets[@]} ]]; then
+           impl=${targets[(${impl} - 1)]}
+       else
+           die -q "Invalid impl ($impl) !"
+       fi
+   fi
    # store the configuration to env.d
    store_config ${EROOT}/etc/env.d/blas/${libdir}/config CURRENT "${impl}"
    # generate ld.so.conf file
```

@heroxbd `eselect blas set 1` such usage has been enabled.

**Pull request closed without merging**